k8s_test_harness/plugin.py

```diff
@@ -26,37 +26,47 @@ def _harness_clean(h: harness.Harness):
         h.cleanup()
 
 
-@pytest.fixture(scope="module")
-def h() -> harness.Harness:
-    LOG.debug("Create harness for %s", config.SUBSTRATE)
-    if config.SUBSTRATE == "local":
-        h = harness.LocalHarness()
-    elif config.SUBSTRATE == "lxd":
-        h = harness.LXDHarness()
-    elif config.SUBSTRATE == "multipass":
-        h = harness.MultipassHarness()
-    elif config.SUBSTRATE == "juju":
-        h = harness.JujuHarness()
-    else:
+def _get_harness() -> harness.Harness:
+    LOG.debug("Initializing harness for %s", config.SUBSTRATE)
+    harness_cls_map = {
+        "local": harness.LocalHarness,
+        "lxd": harness.LXDHarness,
+        "multipass": harness.MultipassHarness,
+        "juju": harness.JujuHarness,
+    }
+    harness_cls = harness_cls_map.get(config.SUBSTRATE)
+    if not harness_cls:
         raise harness.HarnessError(
             "TEST_SUBSTRATE must be one of: local, lxd, multipass, juju"
         )
 
+    return harness_cls()
+
+
+@pytest.fixture(scope="module")
+def h_module() -> harness.Harness:
+    h = _get_harness()
+
     yield h
 
     _harness_clean(h)
 
 
-@pytest.fixture(scope="module")
-def module_instance(
-    h: harness.Harness, tmp_path_factory: pytest.TempPathFactory, request
-) -> Generator[harness.Instance, None, None]:
+@pytest.fixture(scope="function")
+def h_function() -> harness.Harness:
+    h = _get_harness()
+    
+    yield h
+
+    _harness_clean(h)
+
+
+def _get_instance(h: harness.Harness, request):
     """Constructs and bootstraps an instance that persists over a test session.
 
     Bootstraps the instance with all k8sd features enabled to reduce testing time.
     """
     LOG.info("Setup node and enable all features")
-
     instance = h.new_instance()
     k8s_util.setup_k8s_snap(instance)
     request.addfinalizer(lambda: k8s_util.purge_k8s_snap(instance))
```

Comment:
btw, this is not necessary (and even not desired). 
The resource is cleaned up by the harness. 

e.g. if you configure SKIP_CLEANUP to inspect the instance then this command will purge the snap (and any logs) before you can inspect it.

Reply:
The problem is that the local harness cleanup is a no-op, so it was failing to bootstrap the second instance as the first one was still there:

https://github.com/canonical/k8s-snap/blob/c56316705b257c470954e8a08e6984f60a9a2ea7/tests/integration/tests/test_util/harness/local.py#L74-L80

Reply:
One idea would be to have a harness property called "supports_cleanup", set it to False for the local harness and do something like:

```
# Some harnesses aren't actually cleaning up instances (e.g. LocalHarness), so
# let's ensure that the k8s instance is properly purged.
if not h.supports_cleanup:
  request.addfinalizer(lambda: k8s_util.purge_k8s_snap(instance))
```

Reply:
Let's rather fix this on the `LocalHarness` level then.
Instead of `self.initialize` we can have a `self.instance` which is none if the instance is not initialized. 
In the cleanup method you then do `k8s_util.purge_k8s(self.instance)`
WDYT?

Reply:
That would also work, wasn't sure if the harness layer is allowed to have any k8s related logic.

Reply:
fair point. let's leave it like this for now.